snuba/utils/streams/batching.py

```diff
@@ -61,6 +62,150 @@ class Offsets:
+                "partition": str(message.partition.index),
+            },
+        )
+
+        # Create the batch only after the first message is seen.
+        if self.__batch is None:
+            self.__batch = Batch()
+
+        result = self.__worker.process_message(message)
+
+        # XXX: ``None`` is indistinguishable from a potentially valid return
+        # value of ``TResult``!
+        if result is not None:
+            self.__batch.results.append(result)
+
+        duration = (time.time() - start) * 1000
+        self.__batch.messages_processed_count += 1
+        self.__batch.processing_time_ms += duration
+        self.__metrics.timing("process_message", duration)
+
+        if message.partition in self.__batch.offsets:
+            self.__batch.offsets[message.partition].hi = message.get_next_offset()
+        else:
+            self.__batch.offsets[message.partition] = Offsets(
+                message.offset, message.get_next_offset()
+            )
+
+    def close(self) -> None:
+        """
+        Close the processor, discarding any messages (without committing
+        offsets) that were previously consumed and processed since the last
+        batch flush.
+        """
+        self.__closed = True
+
+    def __flush(self) -> None:
+        """
+        Flush the active batch and reset the batch state.
+        """
+        assert not self.__closed
+        assert self.__batch is not None, "cannot flush without active batch"
+
+        logger.info(
+            "Flushing %s items (from %r)",
+            len(self.__batch.results),
+            self.__batch.offsets,
+        )
+
+        self.__metrics.timing(
+            "process_message.normalized",
+            self.__batch.processing_time_ms / self.__batch.messages_processed_count,
+        )
+
+        batch_results_length = len(self.__batch.results)
+        if batch_results_length > 0:
+            logger.debug("Flushing batch via worker")
+            flush_start = time.time()
+            self.__worker.flush_batch(self.__batch.results)
+            flush_duration = (time.time() - flush_start) * 1000
+            logger.info("Worker flush took %dms", flush_duration)
+            self.__metrics.timing("batch.flush", flush_duration)
+            self.__metrics.timing(
+                "batch.flush.normalized", flush_duration / batch_results_length
+            )
+
+        logger.debug("Committing offsets for batch")
+        commit_start = time.time()
+        self.__consumer.stage_offsets(
+            {
+                partition: offsets.hi
+                for partition, offsets in self.__batch.offsets.items()
+            }
+        )
+        offsets = self.__consumer.commit_offsets()
+        logger.debug("Committed offsets: %s", offsets)
+        commit_duration = (time.time() - commit_start) * 1000
+        logger.debug("Offset commit took %dms", commit_duration)
+
+        self.__batch = None
+
+
 class BatchingConsumer(Generic[TPayload]):
```

Comment:
I don't have any plans to backport this in the immediate future (especially given that Sentry itself doesn't use it right now.)

I think it'd be nice to eventually pull out `snuba.utils.streams` for reuse but I don't see that happening in the immediate future — the biggest motivator that I could think of for that would be when Sentry is on Python 3 and we could run the post-processing forwarder code off the improved synchronized consumer implementation and potentially use the multiprocessing code as well (since the bottleneck is just going to move there anyway as our consumers get faster.)